dev/archery/archery/bot.py

```diff
@@ -0,0 +1,286 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import operator
+import shlex
+from pathlib import Path
+from functools import partial
+
+import click
+import github
+from ruamel.yaml import YAML
+
+from .utils.crossbow import Crossbow
+from .utils.git import Git
+
+
+class EventError(Exception):
+    pass
+
+
+class CommandError(Exception):
+
+    def __init__(self, message):
+        self.message = message
+
+
+class _CommandMixin:
+
+    def get_help_option(self, ctx):
+        def show_help(ctx, param, value):
+            if value and not ctx.resilient_parsing:
+                raise click.UsageError(ctx.get_help())
+        option = super().get_help_option(ctx)
+        option.callback = show_help
+        return option
+
+    def __call__(self, message, **kwargs):
+        args = shlex.split(message)
+        try:
+            with self.make_context(self.name, args=args, obj=kwargs) as ctx:
+                return self.invoke(ctx)
+        except click.ClickException as e:
+            raise CommandError(e.format_message())
+
+
+class Command(_CommandMixin, click.Command):
+    pass
+
+
+class Group(_CommandMixin, click.Group):
+
+    def command(self, *args, **kwargs):
+        kwargs.setdefault('cls', Command)
+        return super().command(*args, **kwargs)
+
+    def group(self, *args, **kwargs):
+        kwargs.setdefault('cls', Group)
+        return super().group(*args, **kwargs)
+
+    def parse_args(self, ctx, args):
+        if not args and self.no_args_is_help and not ctx.resilient_parsing:
+            raise click.UsageError(ctx.get_help())
+        return super().parse_args(ctx, args)
+
+
+command = partial(click.command, cls=Command)
+group = partial(click.group, cls=Group)
+
+
+class CrossbowCommentFormatter:
```

Comment:
@wesm I copied parts of the implementation from ursabot, but None of the copied code is GPL contaminated.

Reply:
Yes that's fine, thank you